DTLS: MTU auto-reduction during handshake

mpg opened the pull request:
Note: this is based on #1939.
**Note to reviewers:** please only review commits starting from "Implement PMTU auto-reduction in handshake". (So far, this is a two-commits PR.)
**Note to gatekeeper:** you probably want to merge #1939 first, and want to merge it independently anyway as it contains commits newer than this PR.

## Description

This implements a very basic version of the last paragraph of section 4.1.1.1 of RFC 6347 by lowering the MTU to a safe value until the end of the handshake if a flight needs to be retransmitted more than once.

This is necessary for example in the following scenario: a server communicates with many clients, some of which have a nice solid connection that allows large datagrams, while others have a restricted link and IP stack that doesn't allow UDP datagrams larger than 512 bytes of payload. It might be undesirable to unconditionally set the MTU to 512 on the server as this would penalize clients with a stronger connection, but then the more restricted client would never be able to complete a handshake, assuming the server's certificate chain is larger than 512 bytes (the maximum fragment length extension is not enough here as it can only restrict the record payload to 512 bytes, but the UDP payload is 13 bytes larger than that for DTLS). That's where the mechanism implemented here comes into play.

## Status
**READY**

## Requires Backporting

NO, this is a new feature.

## Migrations

No API change or addition.

No ABI change: the structure `ssl_handshake_params` is changed (new field added) but this is only an internal, no application should ever hold objects of that type.

hanno-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -307,6 +307,7 @@ struct mbedtls_ssl_handshake_params
                                               resending messages             */
     unsigned char alt_out_ctr[8];       /*!<  Alternative record epoch/counter
                                               for resending messages         */
+    uint16_t mtu;                       /*!<  Handshake mtu, used to fragment outoing messages */
```
Typo

k-stachowiak commented on library/ssl_cli.c:
```diff
@@ -2721,7 +2730,7 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
      * regardless of the actual code path.
      */
-    if( ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 3 + n )
+    if( ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n )
```
I'm not certain, which PR this change belongs to, so I will repeat my comment from another one: I think the comment above must be updated to reflect the new computation.

mpg replied:
@k-stachowiak Can you clarify how you think the comment should be adapted? At first glance it looks correct to me, but maybe I'm missing something. Both code paths will read `buf[mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n]` so if `ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n` we're in trouble and otherwise we're fine (note the operator is `<=` not `<`).  I think that's what the comment says.

Btw, this commit belongs to #1955 - so if you have further comments on it, please make them on that PR's page so they can be addressed. Thanks!

mpg commented:
@hanno-arm @k-stachowiak Thanks for your reviews! I believe I've addressed all of your feedback, either by changing the code, or explaining why not. Please review again.

mpg commented:
@hanno-arm @k-stachowiak FYI as this is a very small PR, I took the liberty of rebasing it on top of latest #1939 for future merging clarity.

mpg commented:
Note to self: @hanno-arm is merging this into #1918 in order to resolve potential conflicts early, so I should refrain from rewriting history here from now on.

sbutcher-arm commented:
retest

mpg commented:
Note: `tests/scripts/all.sh -k -r -m` passed succesfuly on this branch (with the latest #1939 merged in, which was a trivial merge).
